In [173]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
import scipy

In [98]:
assessments=pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-y-baraulia/Statistica/assessments.csv')

## assessments.csv - the file contains information about assessments in the test. Typically, each subject in a semester includes a series of tests with grades, followed by a final exam.

code_module - identification code of the subject.

code_presentation - semester (Identification code).

id_assessment - test (Identification number of the assessment).

assessment_type - type of test. There are three types of assessment: Tutor Marked Assessment (TMA), Computer Marked Assessment (CMA), Exam.

date - information about the final test submission date. Calculated as the number of days from the start of the semester. The start date of the semester is assigned the number 0 (zero).

weight - the weight of the test in percentage in the overall course grade. Exams are usually considered separately and have a weight of 100%; the sum of all other assessments adds up to 100%.






In [100]:
assessments.head()

code_module code_presentation  id_assessment assessment_type   date  weight
0         AAA             2013J           1752             TMA   19.0    10.0
1         AAA             2013J           1753             TMA   54.0    20.0
2         AAA             2013J           1754             TMA  117.0    20.0
3         AAA             2013J           1755             TMA  166.0    20.0
4         AAA             2013J           1756             TMA  215.0    30.0

DataFrame size.

In [101]:

assessments.shape

(206, 6)

Data type.

In [102]:
assessments.dtypes

code_module           object
code_presentation     object
id_assessment          int64
assessment_type       object
date                 float64
weight               float64
dtype: object

Descriptive statistics.

In [103]:
assessments.describe(include='all')

code_module code_presentation  id_assessment assessment_type  \
count          206               206     206.000000             206   
unique           7                 4            NaN               3   
top            FFF             2014B            NaN             TMA   
freq            52                57            NaN             106   
mean           NaN               NaN   26473.975728             NaN   
std            NaN               NaN   10098.625521             NaN   
min            NaN               NaN    1752.000000             NaN   
25%            NaN               NaN   15023.250000             NaN   
50%            NaN               NaN   25364.500000             NaN   
75%            NaN               NaN   34891.750000             NaN   
max            NaN               NaN   40088.000000             NaN   

              date      weight  
count   195.000000  206.000000  
unique         NaN         NaN  
top            NaN         NaN  
freq           NaN         NaN  
mean    145.005128   20.873786  
std      76.001119   30.384224  
min      12.000000    0.000000  
25%      71.000000    0.000000  
50%     152.000000   12.500000  
75%     222.000000   24.250000  
max     261.000000  100.000000

Checking for missing data.

In [104]:
assessments.isna().sum()

code_module           0
code_presentation     0
id_assessment         0
assessment_type       0
date                 11
weight                0
dtype: int64

In [105]:
courses=pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-y-baraulia/Statistica/courses.csv')

# courses.csv — the file contains a list of subjects by semesters.

code_module — subject (Identification code).

code_presentation — semester (Identification code).

module_presentation_length — duration of the semester in days.

In [107]:
courses.head()


code_module code_presentation  module_presentation_length
0         AAA             2013J                         268
1         AAA             2014J                         269
2         BBB             2013J                         268
3         BBB             2014J                         262
4         BBB             2013B                         240

DataFrame size.

In [108]:
courses.shape

(22, 3)

Data type.

In [109]:
courses.dtypes

code_module                   object
code_presentation             object
module_presentation_length     int64
dtype: object

Descriptive statistics.

In [110]:
courses.describe(include='all')

code_module code_presentation  module_presentation_length
count           22                22                   22.000000
unique           7                 4                         NaN
top            BBB             2014J                         NaN
freq             4                 7                         NaN
mean           NaN               NaN                  255.545455
std            NaN               NaN                   13.654677
min            NaN               NaN                  234.000000
25%            NaN               NaN                  241.000000
50%            NaN               NaN                  261.500000
75%            NaN               NaN                  268.000000
max            NaN               NaN                  269.000000

Checking for missing data.

In [111]:
courses.isna().sum()

code_module                   0
code_presentation             0
module_presentation_length    0
dtype: int64

In [112]:
student_Assessment=pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-y-baraulia/Statistica/studentAssessment.csv')

## studentAssessment.csv — this file contains the results of student assessments. If a student does not submit an assignment for evaluation, the result is not recorded in the table.
id_assessment — test (Identification number).

id_student — student's identification number.

date_submitted — date of the student's test submission, measured as the number of days from the start of the semester.

is_banked — the fact of transferring the test from the previous semester (sometimes courses are credited to students returning from academic leave).

score — the student's score on this test. The range is from 0 to 100. A score below 40 is considered a failed/unsuccessful test.

In [114]:
student_Assessment.head()

id_assessment  id_student  date_submitted  is_banked  score
0           1752       11391              18          0   78.0
1           1752       28400              22          0   70.0
2           1752       31604              17          0   72.0
3           1752       32885              26          0   69.0
4           1752       38053              19          0   79.0

DataFrame size.

In [115]:
student_Assessment.shape

(173912, 5)

Data type.

In [116]:
student_Assessment.dtypes

id_assessment       int64
id_student          int64
date_submitted      int64
is_banked           int64
score             float64
dtype: object

Descriptive statistics.

In [117]:
student_Assessment.describe(include='all')

id_assessment    id_student  date_submitted      is_banked  \
count  173912.000000  1.739120e+05   173912.000000  173912.000000   
mean    26553.803556  7.051507e+05      116.032942       0.010977   
std      8829.784254  5.523952e+05       71.484148       0.104194   
min      1752.000000  6.516000e+03      -11.000000       0.000000   
25%     15022.000000  5.044290e+05       51.000000       0.000000   
50%     25359.000000  5.852080e+05      116.000000       0.000000   
75%     34883.000000  6.344980e+05      173.000000       0.000000   
max     37443.000000  2.698588e+06      608.000000       1.000000   

               score  
count  173739.000000  
mean       75.799573  
std        18.798107  
min         0.000000  
25%        65.000000  
50%        80.000000  
75%        90.000000  
max       100.000000

Checking for missing data.

In [118]:
student_Assessment.isna().sum()

id_assessment       0
id_student          0
date_submitted      0
is_banked           0
score             173
dtype: int64

In [119]:
student_Registration=pd.read_csv('/mnt/HC_Volume_18315164/home-jupyter/jupyter-y-baraulia/Statistica/studentRegistration.csv')

## studentRegistration.csv — this file contains information about the time when a student registered to take a course in a semester.
code_module — subject (Identification code).

code_presentation — semester (Identification code).

id_student — student's identification number.

date_registration — date of the student's registration. This is the number of days measured from the start of the semester (for example, a negative value of -30 means that the student registered for the course 30 days before its start).

date_unregistration — date of the student's unregistration from the subject. For students who have completed the course, this field remains empty.

In [121]:
student_Registration.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                  NaN  
1                  NaN  
2                 12.0  
3                  NaN  
4                  NaN

DataFrame size.

In [122]:
student_Registration.shape

(32593, 5)

Data type.

In [123]:
student_Registration.dtypes

code_module             object
code_presentation       object
id_student               int64
date_registration      float64
date_unregistration    float64
dtype: object

Descriptive statistics.

In [124]:
student_Registration.describe(include='all')

code_module code_presentation    id_student  date_registration  \
count        32593             32593  3.259300e+04       32548.000000   
unique           7                 4           NaN                NaN   
top            BBB             2014J           NaN                NaN   
freq          7909             11260           NaN                NaN   
mean           NaN               NaN  7.066877e+05         -69.411300   
std            NaN               NaN  5.491673e+05          49.260522   
min            NaN               NaN  3.733000e+03        -322.000000   
25%            NaN               NaN  5.085730e+05        -100.000000   
50%            NaN               NaN  5.903100e+05         -57.000000   
75%            NaN               NaN  6.444530e+05         -29.000000   
max            NaN               NaN  2.716795e+06         167.000000   

        date_unregistration  
count          10072.000000  
unique                  NaN  
top                     NaN  
freq                    NaN  
mean              49.757645  
std               82.460890  
min             -365.000000  
25%               -2.000000  
50%               27.000000  
75%              109.000000  
max              444.000000

Checking for missing data.

In [125]:
student_Registration.isna().sum()

code_module                0
code_presentation          0
id_student                 0
date_registration         45
date_unregistration    22521
dtype: int64

## Task 1 
## How many students successfully passed only one course?

To complete this task, it is necessary to merge two dataframes, namely student_Assessment and assessments.

In [128]:
score_exam=student_Assessment.merge(assessments.query('assessment_type=="Exam"'), how='inner', on='id_assessment')
score_exam

id_assessment  id_student  date_submitted  is_banked  score code_module  \
0             24290      558914             230          0   32.0         CCC   
1             24290      559706             234          0   78.0         CCC   
2             24290      559770             230          0   54.0         CCC   
3             24290      560114             230          0   64.0         CCC   
4             24290      560311             234          0  100.0         CCC   
...             ...         ...             ...        ...    ...         ...   
4954          25368      652680             243          0   91.0         DDD   
4955          25368      652732             249          0   47.0         DDD   
4956          25368      652965             242          0   87.0         DDD   
4957          25368      653051             242          0   27.0         DDD   
4958          25368      653319             250          0   64.0         DDD   

     code_presentation assessment_type  date  weight  
0                2014B            Exam   NaN   100.0  
1                2014B            Exam   NaN   100.0  
2                2014B            Exam   NaN   100.0  
3                2014B            Exam   NaN   100.0  
4                2014B            Exam   NaN   100.0  
...                ...             ...   ...     ...  
4954             2014J            Exam   NaN   100.0  
4955             2014J            Exam   NaN   100.0  
4956             2014J            Exam   NaN   100.0  
4957             2014J            Exam   NaN   100.0  
4958             2014J            Exam   NaN   100.0  

[4959 rows x 10 columns]

In the obtained dataframe, you can see information filtered by the type of test. In our case, we needed data on exam results.

In [130]:
score_40=score_exam.query('score>=40')
score_40

id_assessment  id_student  date_submitted  is_banked  score code_module  \
1             24290      559706             234          0   78.0         CCC   
2             24290      559770             230          0   54.0         CCC   
3             24290      560114             230          0   64.0         CCC   
4             24290      560311             234          0  100.0         CCC   
5             24290      560494             230          0   92.0         CCC   
...             ...         ...             ...        ...    ...         ...   
4952          25368      652617             243          0   56.0         DDD   
4954          25368      652680             243          0   91.0         DDD   
4955          25368      652732             249          0   47.0         DDD   
4956          25368      652965             242          0   87.0         DDD   
4958          25368      653319             250          0   64.0         DDD   

     code_presentation assessment_type  date  weight  
1                2014B            Exam   NaN   100.0  
2                2014B            Exam   NaN   100.0  
3                2014B            Exam   NaN   100.0  
4                2014B            Exam   NaN   100.0  
5                2014B            Exam   NaN   100.0  
...                ...             ...   ...     ...  
4952             2014J            Exam   NaN   100.0  
4954             2014J            Exam   NaN   100.0  
4955             2014J            Exam   NaN   100.0  
4956             2014J            Exam   NaN   100.0  
4958             2014J            Exam   NaN   100.0  

[4392 rows x 10 columns]

In [131]:
score_40.groupby('id_student').agg({'code_module':'nunique'}).query('code_module==1')

code_module
id_student             
23698                 1
24213                 1
27116                 1
28046                 1
28787                 1
...                 ...
2694886               1
2694933               1
2695608               1
2697181               1
2698251               1

[3805 rows x 1 columns]

### We conclude that a total of 3805 students successfully passed at least one exam (scored 40 or higher).

## Task 2
## Identify the most difficult and the easiest exams: find courses and exams within the course that have the lowest and highest completion rates.
### *Completion rate = number of successful exams / total number of attempts to pass the exam.

In [134]:
(score_40.id_assessment.value_counts()/score_exam.id_assessment.value_counts()).sort_values(ascending=False)

25361    0.925573
25354    0.907025
24290    0.888889
25368    0.886316
24299    0.872432
25340    0.837209
Name: id_assessment, dtype: float64

In [135]:
assessments.query('id_assessment==25361')

code_module code_presentation  id_assessment assessment_type   date  \
101         DDD             2014B          25361            Exam  241.0   

     weight  
101   100.0

In [136]:
assessments.query('id_assessment==25340')

code_module code_presentation  id_assessment assessment_type   date  weight
87         DDD             2013B          25340            Exam  240.0   100.0

## Conclusion:
### The course with the highest completion rate is DDD-2014B;
### The course with the lowest completion rate is DDD-2013B.

## Task 3
## For each subject, determine the average time to pass exams (by passing, we mean the student's last successful completion of the exam).

In [139]:
score_40.code_module.value_counts()

DDD    2709
CCC    1683
Name: code_module, dtype: int64

Only two subjects were successfully passed.


Calculated the maximum and minimum time to pass.

In [142]:
score_40.groupby('code_module').agg({'date_submitted':'max'})

date_submitted
code_module                
CCC                     285
DDD                     279

In [143]:
score_40.groupby('code_module').agg({'date_submitted':'min'})

date_submitted
code_module                
CCC                     230
DDD                     229


Obtained the average time to pass the exam for each subject.

In [144]:

score_40.groupby('code_module').agg({'date_submitted':'mean'})

date_submitted
code_module                
CCC              239.345217
DDD              237.976006

## Task 4 
## Identify the most popular subjects (TOP-3) based on the number of registrations. Also, identify the subjects with the highest dropout rates (TOP-3).

## To solve this task, we can use the student_Registration dataframe, as it contains information about the registration and unregistration times for the courses.

With this code, we can see the total number of times students registered for a course.

In [147]:
student_Registration.groupby('code_module').agg({'date_registration':'count'}).sort_values('date_registration',ascending=False)

date_registration
code_module                   
BBB                       7900
FFF                       7751
DDD                       6257
CCC                       4426
EEE                       2932
GGG                       2534
AAA                        748

With this code, we can see the total number of times students unregistered from a course.

In [176]:
student_Registration.groupby('code_module').agg({'date_unregistration':'count'}).sort_values('date_unregistration',ascending=False)

date_unregistration
code_module                     
FFF                         2380
BBB                         2377
DDD                         2235
CCC                         1947
EEE                          718
GGG                          289
AAA                          126

 And keep in mind that students could register for the course twice in different semesters.
nunique() returns a DataFrame with the count of unique values, excluding NaN and duplicates.

In [151]:
st_rg=student_Registration.groupby('code_module', as_index=False).nunique()
st_rg

code_module  code_presentation  id_student  date_registration  \
0         AAA                  2         712                156   
1         BBB                  4        7692                253   
2         CCC                  2        4251                257   
3         DDD                  4        5848                249   
4         EEE                  3        2859                204   
5         FFF                  4        7397                243   
6         GGG                  3        2525                158   

   date_unregistration  
0                   98  
1                  337  
2                  333  
3                  339  
4                  234  
5                  338  
6                  165

TOP-3 most popular subjects - CCC, BBB, DDD

In [152]:
st_rg.sort_values('date_registration', ascending=False).head(3)

code_module  code_presentation  id_student  date_registration  \
2         CCC                  2        4251                257   
1         BBB                  4        7692                253   
3         DDD                  4        5848                249   

   date_unregistration  
2                  333  
1                  337  
3                  339


TOP-3 with the highest dropout rate - DDD, FFF, DDD

In [153]:
st_rg.sort_values('date_unregistration', ascending=False).head(3)

code_module  code_presentation  id_student  date_registration  \
3         DDD                  4        5848                249   
5         FFF                  4        7397                243   
1         BBB                  4        7692                253   

   date_unregistration  
3                  339  
5                  338  
1                  337

## Task 5
## Using pandas, identify the semester with the lowest course completion rate and the semesters with the longest average course completion times in the period from the beginning of 2013 to the end of 2014.

In [155]:
(score_40.groupby('code_presentation').id_assessment.count()/score_exam.groupby('code_presentation').id_assessment.count()).sort_values(ascending=False)

code_presentation
2013J    0.907025
2014B    0.904013
2014J    0.878659
2013B    0.837209
Name: id_assessment, dtype: float64

In [156]:
score_40.groupby('code_presentation').agg({'date_submitted':'mean'})

date_submitted
code_presentation                
2013B                  230.164683
2013J                  239.509112
2014B                  232.997389
2014J                  243.680279

### Semester with the lowest completion rate: 2013B
### Semester with the longest average course completion times: 2014J

## Task 6
## Often, for a qualitative analysis of the audience, I use approaches based on segmentation.

### Using Python, create adapted RFM clusters of students to qualitatively assess your audience.
### In the adapted clustering, you can choose the following metrics: R - average time to pass one exam, F - course completion rate, M - average score obtained in the exam.
### Describe in detail how you created the clusters.
### For each RFM segment, establish boundaries for the recency, frequency, and monetary metrics for interpreting these clusters.

 Calculating Recency - average time to pass one exam.

In [182]:
Recency =score_exam.groupby('id_student')\
    .agg({'date_submitted':'mean'})\
    .sort_values('date_submitted', ascending = False)
Recency

date_submitted
id_student                
555498               285.0
2341830              279.0
555008               266.0
352637               259.0
695916               256.0
...                    ...
546953               229.0
86047                229.0
2348111              229.0
546830               229.0
548069               229.0

[4633 rows x 1 columns]

 Calculating Frequency - course completion rate.

In [181]:
attempts = score_exam.groupby('id_student')\
    .agg({'date_submitted':'count'})\
    .sort_values('date_submitted', ascending = False)
successfull_attempt = score_40.groupby('id_student')\
    .agg({'date_submitted':'count'})\
    .sort_values('date_submitted', ascending = False)
Frequency = attempts.merge(successfull_attempt, on = 'id_student', how = 'left')
Frequency['success_rate'] = Frequency['date_submitted_y'] / Frequency['date_submitted_x']
Frequency = Frequency[['success_rate']]
Frequency.fillna(0)
Frequency=Frequency.dropna()
Frequency

success_rate
id_student              
588482               1.0
507344               1.0
509102               1.0
645792               1.0
512259               1.0
...                  ...
547445               1.0
547448               1.0
547612               1.0
547737               1.0
2698251              1.0

[4097 rows x 1 columns]

Calculating Monetary - average number of points.

In [179]:
Monetary = score_exam.groupby('id_student')\
    .agg({'score':'mean'})\
    .sort_values('score', ascending = False)
Monetary

score
id_student       
652902      100.0
631473      100.0
588274      100.0
627401      100.0
2131908     100.0
...           ...
338034        7.0
2588543       4.0
556660        0.0
131152        0.0
81351         0.0

[4633 rows x 1 columns]

In [183]:
rfm = Recency.merge(Frequency, on = 'id_student', how = 'outer')
rfm = rfm.merge(Monetary, on = 'id_student')
rfm = rfm.rename(columns={"date_submitted": "Recency", "success_rate": "Frequency", "score":'Monetary'})
rfm

Recency  Frequency  Monetary
id_student                              
555498        285.0        1.0      58.0
2341830       279.0        1.0      49.0
555008        266.0        1.0      56.0
352637        259.0        1.0      96.0
695916        256.0        1.0      67.0
...             ...        ...       ...
546953        229.0        1.0      56.0
86047         229.0        1.0      42.0
2348111       229.0        1.0      62.0
546830        229.0        NaN      33.0
548069        229.0        1.0      78.0

[4633 rows x 3 columns]


Let's calculate the quantiles for each metric.

In [184]:
quintiles = rfm[['Recency', 'Frequency', 'Monetary']].quantile([.2, .4, .6, .8]).to_dict()
quintiles

{'Recency': {0.2: 231.0, 0.4: 237.0, 0.6: 242.0, 0.8: 243.5},
 'Frequency': {0.2: 1.0, 0.4: 1.0, 0.6: 1.0, 0.8: 1.0},
 'Monetary': {0.2: 46.200000000000045, 0.4: 60.0, 0.6: 72.0, 0.8: 84.0}}

In [185]:
quintiles = rfm[['Recency', 'Frequency', 'Monetary']].quantile([.2, .4, .6, .8]).to_dict()
quintiles
def r_score(x):
    if x <= quintiles['Recency'][.2]:
        return 5
    elif x <= quintiles['Recency'][.4]:
        return 4
    elif x <= quintiles['Recency'][.6]:
        return 3
    elif x <= quintiles['Recency'][.8]:
        return 2
    else:
        return 1

def fm_score(x, c):
    if x <= quintiles[c][.2]:
        return 1
    elif x <= quintiles[c][.4]:
        return 2
    elif x <= quintiles[c][.6]:
        return 3
    elif x <= quintiles[c][.8]:
        return 4
    else:
        return 5

In [186]:
rfm['R'] = rfm['Recency'].apply(lambda x: r_score(x))
rfm['F'] = rfm['Frequency'].apply(lambda x: fm_score(x, 'Frequency'))
rfm['M'] = rfm['Monetary'].apply(lambda x: fm_score(x, 'Monetary'))

In [187]:
rfm['RFM Score'] = rfm['R'].map(str) + rfm['F'].map(str) + rfm['M'].map(str)
rfm = rfm.sort_values('RFM Score')
rfm

Recency  Frequency  Monetary  R  F  M RFM Score
id_student                                                 
693182        244.0        1.0      46.0  1  1  1       111
678419        245.0        1.0      44.0  1  1  1       111
685950        244.0        1.0      44.0  1  1  1       111
2464386       244.0        1.0      46.0  1  1  1       111
131933        244.0        1.0      46.0  1  1  1       111
...             ...        ...       ... .. .. ..       ...
614477        230.0        NaN      34.0  5  5  1       551
478550        230.0        NaN      34.0  5  5  1       551
547986        230.0        NaN      33.0  5  5  1       551
551604        230.0        NaN      14.0  5  5  1       551
438662        231.0        NaN      38.0  5  5  1       551

[4633 rows x 7 columns]

In [189]:
 segments = {
    r'[4-5][1-2]': 'zubrilki',
    r'[1-2][4-5]': 'poslednie_ryadi',
    r'[4-5]5': 'vunderkindi',
    r'[4-5][3-4]': 'horoshisti',
    r'35': 'uspeshnie',
    r'3[3-4]': 'srednyachki',
    r'3[1-2]': 'nizhe_srednego',
    r'[1-2][1-2]': 'dvoechniki',
    r'[1-2]3': 'tusovchiki'
}

rfm['Segment'] = rfm['M'].map(str) + rfm['R'].map(str)
rfm['Segment'] = rfm['Segment'].replace(segments, regex=True)
rfm.head()

Recency  Frequency  Monetary  R  F  M RFM Score     Segment
id_student                                                             
693182        244.0        1.0      46.0  1  1  1       111  dvoechniki
678419        245.0        1.0      44.0  1  1  1       111  dvoechniki
685950        244.0        1.0      44.0  1  1  1       111  dvoechniki
2464386       244.0        1.0      46.0  1  1  1       111  dvoechniki
131933        244.0        1.0      46.0  1  1  1       111  dvoechniki

## Here are the 9 segments::
1) zubrilki - spend a lot of time and perform well in exams.;

2) poslednie_ryadi - spend little time and perform poorly in exams;

3) vunderkindi - spend little time and still perform well in exams;

4) horoshisti  - spend enough time and achieve high grades;

5) uspeshnie - spend little time and get average grades;

6) srednyachki - spend an average amount of time and receive average grades;

7) nizhe_srednego - spend a lot of time but achieve not very high grades, perhaps needing more time or assistance;

8) dvoechniki - spend a lot of time but receive low grades (or cram at the last moment without spending much time);

9) tusovchiki - receive low grades and could dedicate more time.